arch: arm: cortex_m: pm_s2ram: add support for ARMv6-M and ARMv7-M

mathieuchopstm opened the pull request:
This PR makes the ARM implementation of suspend-to-RAM compatible with the ARMv6-M and ARMv7-M architectures.

Tested on Cortex-M0+ (ARMv6-M) with a minimal example that invokes `arch_pm_s2ram_suspend` directly; CPU registers appear to be properly restored after resuming from the low-power mode.

Not tested on ARMv7-M due to lack of hardware.

mathieuchopstm commented:
Rebased on main to trigger CI again (failure was due to faulty code in `main`).

HesselM commented on arch/arm/core/cortex_m/pm_s2ram.S:
```diff
@@ -20,131 +20,198 @@ GTEXT(pm_s2ram_mark_set)
+	push	{ r1-r5, lr }
+#else
+	/* ARMv7-M and ARMv8-M have no limitation on `push` */
+	push	{ r5-r12 }
```
Should this not be "push {r4-r12, lr}"?

HesselM commented on arch/arm/core/cortex_m/pm_s2ram.S:
```diff
@@ -20,131 +20,198 @@ GTEXT(pm_s2ram_mark_set)
+	pop	{ r4-r7 }
+#else
+	/* ARMv7-M and ARMv8-M have no limitation on `push` */
+	push	{ r5-r12 }
```
This should be a "pop" instead of push.
And probably "pop {r4-r12, lr}"

HesselM commented on include/zephyr/arch/arm/cortex_m/cpu.h:
```diff
@@ -52,18 +52,25 @@ extern "C" {
+	uint32_t msplim;
+	uint32_t psplim;
+#		endif
+#	endif
```
The ifdef/endif alignment is incorrect.

mathieuchopstm replied:
How so? Note that the `#endif` for `#ifdef CONFIG_PM_S2RAM` is later down the file.

HesselM replied:
Sorry, i meant the white-spacing. According to the code-styling rules it is my understanding that every preprocessing-directive start at the beginning of the line and should not have whitespaces between the hashtag and the tag.

HesselM replied:
I did a search, and to my surprise could not find anything about this (I did not even know this was possible). Both zephyr and linux style-guide do not mention it and the clangd parser also does not complain. I can even see in some libraries the same usage:

```
#if OPENSSL_VERSION_NUMBER >= 0x10100000L /* 1.1.0 */
      SSL_CTX_set_min_proto_version(backend->ctx, SSL3_VERSION);
      SSL_CTX_set_max_proto_version(backend->ctx, SSL3_VERSION);
#else
      ctx_options |= SSL_OP_NO_SSLv2;
      ctx_options |= SSL_OP_NO_TLSv1;
#  if OPENSSL_VERSION_NUMBER >= 0x1000100FL
      ctx_options |= SSL_OP_NO_TLSv1_1;
      ctx_options |= SSL_OP_NO_TLSv1_2;
#    ifdef TLS1_3_VERSION
      ctx_options |= SSL_OP_NO_TLSv1_3;
#    endif
#  endif
#endif
```

But: i could not find this setup in the zephyr kernel itself, and the usage in the libraries / external sources shows much shorter indents: 1 or 2 spaces per indent instead of 4. With only 1 or 2 spaces indent, the directive is still visually (easy) detectable as directive. With the 4-spaces the directive is much harder to spot between the lines of code.

So, as I'm not part of the zephyr-team, i cannot say a lot about it, but my guess is that - as I could not find this approach in the kernel - the zephyr-team does not use it. And if it is used, the indent is only 1-2 spaces to keep the directive readable / easy recognisable.

HesselM commented:
I've done a very (limited) test on armv7-M (m4) and it seems to work. I can't do a full test as we are still working on fully implementing s2ram for our platform, which is the qn9090 soc from NXP.

ithinuel commented:
I apologize for the delay in providing a review. I am not familiar with all the registers saved/restored in this module.

Some of the things that confuse me a bit are for example saving & restoring EPSR which cannot be accessed by MSR/MRS, and IPSR which is read only.

mathieuchopstm commented:
> I apologize for the delay in providing a review. I am not familiar with all the registers saved/restored in this module.

No problem :smile:

> Some of the things that confuse me a bit are for example saving & restoring EPSR which cannot be accessed by MSR/MRS, and IPSR which is read only.

I agree with you that some things performed by this module appear incorrect or useless, but I kept it as-is to not introduce functional changes. I'm fine with removing them though.

I am preparing a reworked version of this PR to make review easier and address the points raised by @HesselM ...

mathieuchopstm commented:
Patch v2 (+ rebase on top of 2590c48d40de8684609436c28b080a693ef56b09):
* split in separate commits to ease review
* address issues pointed out by @HesselM

ithinuel commented on arch/arm/core/cortex_m/pm_s2ram.S:
```diff
@@ -14,58 +14,132 @@
+	SAVE_SPECIAL_REG(msp,		\cpu_ctx, \tmp_reg)
+	SAVE_SPECIAL_REG(psp,		\cpu_ctx, \tmp_reg)
+	SAVE_SPECIAL_REG(apsr,		\cpu_ctx, \tmp_reg)
+	SAVE_SPECIAL_REG(ipsr,		\cpu_ctx, \tmp_reg)
+	SAVE_SPECIAL_REG(epsr,		\cpu_ctx, \tmp_reg)
```
EPSR reads as 0 and writes are ignored.
IPSR can be read but writes are ignored too.
See B1.4.2 of Armv6-m Architecture Reference Manual and B3.5 of the Armv8-m Architecture Reference Manual.

So trying to save these registers has no effect. These should be removed.

ithinuel commented on arch/arm/core/cortex_m/pm_s2ram.S:
```diff
@@ -91,7 +165,7 @@ SECTION_FUNC(TEXT, arch_pm_s2ram_suspend)
 	mov	r0, r4
 
-	pop	{r4-r12, lr}
+	pop_gprs
```
IIUC at this stage, the registers at the top of the stack are the special purpose ones.
`sp` should be shifted first before restoring `gprs`.

mathieuchopstm replied:
Special purpose registers are saved in the `_cpu_context` variable and not on the stack, AFAIU?

ithinuel replied:
That’s right! my mistake.

ithinuel commented on arch/arm/core/offsets/offsets_aarch32.c:
```diff
@@ -83,17 +83,24 @@ GEN_OFFSET_SYM(_thread_stack_info_t, start);
+	/* Registers present on ARMv7-M and ARMv8-M */
+	GEN_OFFSET_SYM(_cpu_context_t, faultmask);
+	GEN_OFFSET_SYM(_cpu_context_t, basepri);
+#ifdef CONFIG_ARMV8_M_MAINLINE
+	/* Registers present only on ARMv8-M */
+	GEN_OFFSET_SYM(_cpu_context_t, msplim);
+	GEN_OFFSET_SYM(_cpu_context_t, psplim);
+#endif /* CONFIG_ARMV8_M_MAINLINE */
```
These are also present in on Armv8-m baseline so support for these should be added too.

ithinuel reviewed (changes requested):
Thank you Mathieu for the updated version.
If you don’t mind addressing the few points listed bellow, I think this will be in a good shape for approval 👍

mathieuchopstm commented on arch/arm/core/cortex_m/pm_s2ram.S:
```diff
@@ -20,131 +20,198 @@ GTEXT(pm_s2ram_mark_set)
+	 * Call the `system_off` function passed as parameter.
+	 * This call will not return if system is suspended.
 	 */
 	blx	r0
```
Already fixed upstream by 2590c48d40de8684609436c28b080a693ef56b09 - PR has been rebased so this should be good.

mathieuchopstm commented:
Patch v3 (rebased on main, just in case...)
* add commit that removes backup/restore of IPSR and EPSR
* backup/restore `Xsplim` registers for ARMv8-M Baseline too (`CONFIG_ARMV8_M_BASELINE`)

ithinuel commented on arch/arm/core/offsets/offsets_aarch32.c:
```diff
@@ -83,17 +83,22 @@ GEN_OFFSET_SYM(_thread_stack_info_t, start);
 GEN_OFFSET_SYM(_cpu_context_t, faultmask);
 GEN_OFFSET_SYM(_cpu_context_t, basepri);
-GEN_OFFSET_SYM(_cpu_context_t, control);
+#if defined(ARMV8_M_BASELINE) || defined(CONFIG_ARMV8_M_MAINLINE)
+/* Registers present only on ARMv8-M */
+GEN_OFFSET_SYM(_cpu_context_t, msplim);
+GEN_OFFSET_SYM(_cpu_context_t, psplim);
+#endif /* CONFIG_ARMV8_M_MAINLINE */
```
`ARMV8_M_BASELINE` cannot be reached here because they are not `CONFIG_ARMV7_M_ARMV8_M_MAINLINE`.

I think this block needs to be outside of `#ifdef CONFIG_ARMV7_M_ARMV8_M_MAINLINE`.

mathieuchopstm replied:
Good catch! Fixing this.

ithinuel reviewed (dismissed):
LGTM. Not actually tested myself on any hardware so we may have overlooked some corner cases. Still, this is already more portable than the current implementation 👍

stephanosio commented on arch/arm/core/cortex_m/pm_s2ram.S:
```diff
@@ -14,58 +14,138 @@
 GTEXT(pm_s2ram_mark_check_and_clear)
 GDATA(_cpu_context)
 
+# Pushes registers r4~r12 and lr on the stack.
+# r0 is unmodified, other GPRs may be overwritten.
```
Nit: Let's try to stick with the comment style used throughout the Zephyr ASM files, which is C.

stephanosio commented on arch/arm/core/cortex_m/pm_s2ram.S:
```diff
@@ -14,58 +14,138 @@
+	/* `push` on ARMv6-M / ARMv8-M Baseline:
+	 * only r0~r7 and lr may be pushed
+	 */
+	push	{ r4-r7 }
```
Nit: Align with the existing style.
```suggestion
	push	{r4-r7}
```

stephanosio commented on arch/arm/core/cortex_m/pm_s2ram.S:
```diff
@@ -47,36 +47,82 @@ GDATA(_cpu_context)
-	SAVE_SPECIAL_REG(control,	\cpu_ctx, \tmp_reg)
+#endif /* CONFIG_ARMV7_M_ARMV8_M_MAINLINE */
+
+#if defined(CONFIG_ARMV8_M_BASELINE) || defined(CONFIG_ARMV8_M_MAINLINE)
+	/* Registers present only on ARMv8-M (Baseline / Mainline) */
+	SAVE_SPECIAL_REG(msplim,	\cpu_ctx, \tmp_reg)
+	SAVE_SPECIAL_REG(psplim,	\cpu_ctx, \tmp_reg)
+#endif /* CONFIG_ARMV8_M_BASELINE || CONFIG_ARMV8_M_MAINLINE */
```
Access to the SPLIM registers should be guarded by `CONFIG_CPU_CORTEX_M_HAS_SPLIM`.

mathieuchopstm committed:
arch: arm: cortex_m: pm_s2ram: ignore xPSR

Remove all xPSR-related registers from struct __cpu_context, and the
associated save/restore code in S2RAM code, as they are not needed:

* EPSR and IPSR are read-only - they cannot be "restored"
* Bits N, V, Z, C, V, Q, and GE (if DSP Extension is implemented) of APSR
  could be restored, but this is not needed as the AAPCS indicates these
  bits to be "undefined on entry to or return from a public interface"

Signed-off-by: Mathieu Choplain <mathieu.choplain@st.com>

mathieuchopstm committed:
arch: arm: cortex_m: pm_s2ram: use macros to access struct __cpu_context

Use macros to wrap the interaction between the assembly code and the
struct __cpu_context. This helps making the assembly more readable.

Signed-off-by: Mathieu Choplain <mathieu.choplain@st.com>

mathieuchopstm committed:
arch: arm: cortex_m: pm_s2ram: wrap context save/restore in macros

Wrap the CPU register save/restore operations (GPR and special registers)
in macros to make core logic simpler to follow. This is also a preparatory
step to introduce ARMv6-M and ARMv7-M support.

Signed-off-by: Mathieu Choplain <mathieu.choplain@st.com>

mathieuchopstm committed:
arch: arm: cortex_m: pm_s2ram: add support for all architectures

Extend the ARM M-profile suspend-to-RAM implementation to be compatible
with all versions of the M-profile supported by Zephyr: ARMv6-M, ARMv7-M,
and ARMv8-M Baseline.

Signed-off-by: Mathieu Choplain <mathieu.choplain@st.com>

mrkhldn commented:
Seems to work on M7
